### PR TITLE
mat: tracer for band types

### PR DESCRIPTION
**Pull request by jonreiter**
Please take a look.  Similar to https://github.com/gonum/gonum/issues/938.

### CODE REVIEW

File: mat/band.go [hunk 1]
```diff
@@ -34,6 +34,9 @@ type Banded interface {
 	// TBand is the equivalent of the T() method in the Matrix
 	// interface but guarantees the transpose is of banded type.
 	TBand() Banded
+
+	// Trace returns the trace
+	Trace() float64
```
Comment (kortschak) [inline mat/band.go]: This shouldn't be here. It will be an optional interface.

File: mat/band.go [hunk 2]
```diff
@@ -101,6 +104,11 @@ func (t TransposeBand) UntransposeBand() Banded {
 	return t.Banded
 }
 
+// Trace returns the Banded field.
```
Comment (kortschak) [inline mat/band.go]: This is not necessary.

File: mat/diagonal.go [hunk 3]
```diff
@@ -60,6 +60,9 @@ type Diagonal interface {
 	// to allow the use of Diagonal types in triangular banded functions.
 	TriBand() (n, k int, kind TriKind)
 	TTriBand() TriBanded
+
```
Comment (kortschak) [inline mat/diagonal.go]: Please remove this.

File: mat/mul_test.go [hunk 4]
```diff
@@ -239,6 +239,10 @@ func (m *basicBanded) TBand() Banded {
 	return TransposeBand{m}
 }
 
+func (m *basicBanded) Trace() float64 {
```
Comment (kortschak) [inline mat/mul_test.go]: Please remove this; the `basicType` types do not satisfy anything except their base method set.

File: mat/mul_test.go [hunk 5]
```diff
@@ -315,6 +319,10 @@ func (m *basicSymBanded) Bandwidth() (kl, ku int) {
 	return (*SymBandDense)(m).Bandwidth()
 }
 
+func (m *basicSymBanded) Trace() float64 {
```
Comment (kortschak) [inline mat/mul_test.go]: Please remove this.

File: mat/mul_test.go [hunk 6]
```diff
@@ -355,6 +363,10 @@ func (m *basicTriBanded) TriBand() (n, k int, kind TriKind) {
 	return (*TriBandDense)(m).TriBand()
 }
 
+func (m *basicTriBanded) Trace() float64 {
```
Comment (kortschak) [inline mat/mul_test.go]: Please remove this.

File: mat/mul_test.go [hunk 7]
```diff
@@ -407,6 +419,10 @@ func (m *basicDiagonal) TriBand() (n, k int, kind TriKind) {
 	return (*DiagDense)(m).TriBand()
 }
 
+func (m *basicDiagonal) Trace() float64 {
```
Comment (kortschak) [inline mat/mul_test.go]: Please remove this.

File: mat/triband.go [hunk 8]
```diff
@@ -147,6 +147,11 @@ func (t TransposeTriBand) UntransposeTriBand() TriBanded {
 	return t.TriBanded
 }
 
+// Trace returns the trace
```
Comment (kortschak) [inline mat/triband.go]: Please remove this.

**Comment by jonreiter** [comment]
I believe all sorted

File: mat/band.go [hunk 9]
```diff
@@ -261,3 +261,17 @@ func (b *BandDense) Zero() {
...
 }
+
+// Trace computes the trace of the matrix.
+func (b *BandDense) Trace() float64 {
+	r, c := b.Dims()
+	if r != c {
+		panic(ErrShape)
+	}
+	rb := b.RawBand()
+	tr := 0.0
```
Comment (kortschak) [inline mat/band.go]: Conventionally we would use `var tr float64` in this case.

File: mat/matrix_test.go [hunk 10]
```diff
@@ -562,6 +562,47 @@ func TestTrace(t *testing.T) {
 	testOneInputFunc(t, "Trace", f, denseComparison, sameAnswerFloat, isAnyType, isSquare)
 }
 
+func TestTracer(t *testing.T) {
+	for _, test := range []struct {
+		a     Tracer
+		trace float64
```
Comment (kortschak) [inline mat/matrix_test.go]: s/trace/want/g

File: mat/matrix_test.go [hunk 11]
```diff
@@ -562,6 +562,47 @@ func TestTrace(t *testing.T) {
...
+		{
+			a:     NewTriBandDense(6, 2, Upper, []float64{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 0, 15, 0, 0}),
+			trace: 50,
+		},
+		{
+			a:     NewTriBandDense(6, 2, Lower, []float64{0, 0, 1, 0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15}),
+			trace: 46,
+		},
+	} {
+		trace := test.a.Trace()
```
Comment (kortschak) [inline mat/matrix_test.go]: s/trace/got/

Comment (jonreiter) [reply mat/matrix_test.go]: changed.

i'll note that TestTrace, which is quite old, does not use the want/got naming.  trying to keep this change as narrow as possible.

File: mat/diagonal.go [hunk 12]
```diff
@@ -309,3 +309,14 @@ func (d *DiagDense) IsZero() bool {
 	// zeros in this case. See comment in Reset().
 	return d.mat.Inc == 0
 }
+
+// Trace returns the trace.
+func (d *DiagDense) Trace() float64 {
+	rb := d.RawBand()
+	tr := 0.0
```
Comment (kortschak) [inline mat/diagonal.go]: Sorry. Throughout.

File: mat/symband.go [hunk 13]
```diff
@@ -219,3 +219,13 @@ func (s *SymBandDense) DoColNonZero(j int, fn func(i, j int, v float64)) {
 		}
 	}
 }
+
+// Trace returns the trace.
+func (s *SymBandDense) Trace() float64 {
+	rb := s.RawSymBand()
+	tr := 0.0
```
Comment (kortschak) [inline mat/symband.go]: .. and here.

File: mat/triband.go [hunk 14]
```diff
@@ -351,3 +351,17 @@ func (t *TriBandDense) DiagView() Diagonal {
 		},
 	}
 }
+
+// Trace returns the trace.
+func (t *TriBandDense) Trace() float64 {
+	rb := t.RawTriBand()
+	tr := 0.0
```
Comment (kortschak) [inline mat/triband.go]: And here for `tr` and `offsetIndex`; `var off int`.

Comment (jonreiter) [reply mat/triband.go]: fair enough my bad.  all resolved now (i hope!).

**Pull request merged**